File: archive/pulumi-py/Pr1165/lib/modules/security.py
"""
Security module for security groups and S3 bucket with encryption
"""

import pulumi
import pulumi_aws as aws
from typing import Dict


def create_security_groups(region: str, vpc_id: pulumi.Output,
                           tags: Dict, provider: aws.Provider) -> Dict:
  """Create security groups with least privilege access"""

  # Web tier security group
  web_sg = aws.ec2.SecurityGroup(
    f"web-sg-{region}",
    name=f"web-sg-{region}",
    description="Security group for web tier",
    vpc_id=vpc_id,
    tags={**tags, "Name": f"web-sg-{region}", "Tier": "Web"},
    opts=pulumi.ResourceOptions(provider=provider)
  )

  # Web tier rules
  aws.ec2.SecurityGroupRule(
    f"web-ingress-http-{region}",
    type="ingress",
    from_port=80,
    to_port=80,
    protocol="tcp",
    cidr_blocks=["0.0.0.0/0"],
    security_group_id=web_sg.id,
    opts=pulumi.ResourceOptions(provider=provider)
  )

  aws.ec2.SecurityGroupRule(
    f"web-ingress-https-{region}",
    type="ingress",
    from_port=443,
    to_port=443,
    protocol="tcp",
    cidr_blocks=["0.0.0.0/0"],
    security_group_id=web_sg.id,
    opts=pulumi.ResourceOptions(provider=provider)
  )

  aws.ec2.SecurityGroupRule(
    f"web-egress-all-{region}",
    type="egress",
    from_port=0,
    to_port=0,
    protocol="-1",
    cidr_blocks=["0.0.0.0/0"],
    security_group_id=web_sg.id,
    opts=pulumi.ResourceOptions(provider=provider)
  )

  # Application tier security group
  app_sg = aws.ec2.SecurityGroup(
    f"app-sg-{region}",
    name=f"app-sg-{region}",
    description="Security group for application tier",
    vpc_id=vpc_id,
    tags={**tags, "Name": f"app-sg-{region}", "Tier": "Application"},
    opts=pulumi.ResourceOptions(provider=provider)
  )

  # App tier rules (only from web tier)
  aws.ec2.SecurityGroupRule(
    f"app-ingress-from-web-{region}",
    type="ingress",
    from_port=8080,
    to_port=8080,
    protocol="tcp",
    source_security_group_id=web_sg.id,
    security_group_id=app_sg.id,
    opts=pulumi.ResourceOptions(provider=provider)
  )

  aws.ec2.SecurityGroupRule(
    f"app-egress-all-{region}",
    type="egress",
    from_port=0,
    to_port=0,
    protocol="-1",
    cidr_blocks=["0.0.0.0/0"],
    security_group_id=app_sg.id,
    opts=pulumi.ResourceOptions(provider=provider)
  )

  # Database tier security group
  db_sg = aws.ec2.SecurityGroup(
    f"db-sg-{region}",
    name=f"db-sg-{region}",
    description="Security group for database tier",
    vpc_id=vpc_id,
    tags={**tags, "Name": f"db-sg-{region}", "Tier": "Database"},
    opts=pulumi.ResourceOptions(provider=provider)
  )

  # Database tier rules (only from app tier)
  aws.ec2.SecurityGroupRule(
    f"db-ingress-from-app-{region}",
    type="ingress",
    from_port=3306,
    to_port=3306,
    protocol="tcp",
    source_security_group_id=app_sg.id,
    security_group_id=db_sg.id,
    opts=pulumi.ResourceOptions(provider=provider)
  )

  return {
    "web_sg": web_sg,
    "app_sg": app_sg,
    "db_sg": db_sg
  }


def create_s3_bucket(region: str, tags: Dict, provider: aws.Provider) -> aws.s3.Bucket:
  """Create S3 bucket with encryption and secure policies"""

  # Create S3 bucket
  bucket = aws.s3.Bucket(
    f"secure-bucket-{region}",
    bucket=f"secure-infrastructure-bucket-{region}-{pulumi.get_stack()}".lower(),
    tags=tags,
    opts=pulumi.ResourceOptions(provider=provider)
  )

  # Enable versioning
  aws.s3.BucketVersioningV2(
    f"bucket-versioning-{region}",
    bucket=bucket.id,
    versioning_configuration=aws.s3.BucketVersioningV2VersioningConfigurationArgs(
      status="Enabled"
    ),
    opts=pulumi.ResourceOptions(provider=provider)
  )

  # Enable server-side encryption
  aws.s3.BucketServerSideEncryptionConfigurationV2(
    f"bucket-encryption-{region}",
    bucket=bucket.id,
    rules=[aws.s3.BucketServerSideEncryptionConfigurationV2RuleArgs(
      apply_server_side_encryption_by_default=aws.s3.BucketServerSideEncryptionConfigurationV2RuleApplyServerSideEncryptionByDefaultArgs(
        sse_algorithm="AES256"
      )
    )],
    opts=pulumi.ResourceOptions(provider=provider)
  )

  # Block public access
  aws.s3.BucketPublicAccessBlock(
    f"bucket-pab-{region}",
    bucket=bucket.id,
    block_public_acls=True,
    block_public_policy=True,
    ignore_public_acls=True,
    restrict_public_buckets=True,
    opts=pulumi.ResourceOptions(provider=provider)
  )

  # Bucket policy to enforce SSL
  bucket_policy = pulumi.Output.all(bucket.arn).apply(
    lambda args: f"""{{
            "Version": "2012-10-17",
            "Statement": [
                {{
                    "Sid": "DenyInsecureConnections",
                    "Effect": "Deny",
                    "Principal": "*",
                    "Action": "s3:*",
                    "Resource": [
                        "{args[0]}",
                        "{args[0]}/*"
                    ],
                    "Condition": {{
                        "Bool": {{
                            "aws:SecureTransport": "false"
                        }}
                    }}
                }}
            ]
        }}"""
  )

  aws.s3.BucketPolicy(
    f"bucket-policy-{region}",
    bucket=bucket.id,
    policy=bucket_policy,
    opts=pulumi.ResourceOptions(provider=provider)
  )

  return bucket

File: archive/cdktf-ts/Pr1323/lib/security.ts
import { SecurityGroup } from '@cdktf/provider-aws/lib/security-group';
import { SecurityGroupRule } from '@cdktf/provider-aws/lib/security-group-rule';
import { Construct } from 'constructs';

interface SecurityProps {
  vpcId: string;
  environment: string;
  region: string;
  allowedCidr: string;
  tags: { [key: string]: string };
}

export class Security extends Construct {
  public readonly webSg: SecurityGroup;
  public readonly appSg: SecurityGroup;
  public readonly dbSg: SecurityGroup;

  constructor(scope: Construct, id: string, props: SecurityProps) {
    super(scope, id);

    // Web SG: Only allow HTTP/HTTPS from known IPs
    this.webSg = new SecurityGroup(this, 'WebSG', {
      vpcId: props.vpcId,
      name: `${props.environment}-${props.region}-web-sg`,
      description: 'Web tier SG',
      tags: {
        ...props.tags,
        Name: `${props.environment}-${props.region}-web-sg`,
        Tier: 'web',
      },
    });

    new SecurityGroupRule(this, 'WebHTTP', {
      type: 'ingress',
      fromPort: 80,
      toPort: 80,
      protocol: 'tcp',
      cidrBlocks: [props.allowedCidr],
      securityGroupId: this.webSg.id,
      description: 'Allow HTTP from known IP',
    });

    new SecurityGroupRule(this, 'WebHTTPS', {
      type: 'ingress',
      fromPort: 443,
      toPort: 443,
      protocol: 'tcp',
      cidrBlocks: [props.allowedCidr],
      securityGroupId: this.webSg.id,
      description: 'Allow HTTPS from known IP',
    });

    // App SG: Only allow traffic from web SG
    this.appSg = new SecurityGroup(this, 'AppSG', {
      vpcId: props.vpcId,
      name: `${props.environment}-${props.region}-app-sg`,
      description: 'App tier SG',
      tags: {
        ...props.tags,
        Name: `${props.environment}-${props.region}-app-sg`,
        Tier: 'app',
      },
    });

    new SecurityGroupRule(this, 'AppIngress', {
      type: 'ingress',
      fromPort: 8080,
      toPort: 8080,
      protocol: 'tcp',
      sourceSecurityGroupId: this.webSg.id,
      securityGroupId: this.appSg.id,
      description: 'Allow traffic from web SG',
    });

    // DB SG: Only allow traffic from app SG
    this.dbSg = new SecurityGroup(this, 'DbSG', {
      vpcId: props.vpcId,
      name: `${props.environment}-${props.region}-db-sg`,
      description: 'DB tier SG',
      tags: {
        ...props.tags,
        Name: `${props.environment}-${props.region}-db-sg`,
        Tier: 'db',
      },
    });

    new SecurityGroupRule(this, 'DbIngress', {
      type: 'ingress',
      fromPort: 3306,
      toPort: 3306,
      protocol: 'tcp',
      sourceSecurityGroupId: this.appSg.id,
      securityGroupId: this.dbSg.id,
      description: 'Allow MySQL from app SG',
    });

    // All SGs: Default deny all other inbound, allow all outbound
    [this.webSg, this.appSg, this.dbSg].forEach((sg, i) => {
      new SecurityGroupRule(this, `Egress${i}`, {
        type: 'egress',
        fromPort: 0,
        toPort: 0,
        protocol: '-1',
        cidrBlocks: ['0.0.0.0/0'],
        securityGroupId: sg.id,
        description: 'Allow all outbound',
      });
    });
  }
}
